### PR TITLE
feat: add output_label and input_prediction

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: vowpalwabbit/learner.h [hunk 1]
```diff
@@ -435,9 +437,9 @@ struct learner
   }
 
   prediction_type_t get_output_prediction_type() { return _output_pred_type; }
-
+  prediction_type_t get_input_prediction_type() { return _input_pred_type; }
+  label_type_t get_output_label_type() { return _output_label_type; }
   label_type_t get_input_label_type() { return _input_label_type; }
```
[the remark this answers is not in the document]
Comment (bassmang) [reply vowpalwabbit/learner.h:451]: Resolved, I put in some docs here: https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Matching-Label-and-Prediction-Types-Between-Reductions

File: vowpalwabbit/learner.h [hunk 2]
```diff
@@ -679,8 +693,8 @@ struct reduction_no_data_learner_builder
 
     set_params_per_weight(1);
     // Default here is passthrough.
-    super::set_prediction_type(base->get_output_prediction_type());
-    super::set_label_type(base->get_input_label_type());
+    super::set_output_prediction_type(base->get_output_prediction_type());
+    super::set_input_label_type(base->get_input_label_type());
```
[the remark this answers is not in the document]
Comment (bassmang) [reply vowpalwabbit/learner.h]: Added

File: vowpalwabbit/example.h [hunk 3]
```diff
@@ -70,6 +70,7 @@ struct polyprediction
   VW::continuous_actions::probability_density_function pdf;  // probability density defined over an action range
   VW::continuous_actions::probability_density_function_value pdf_value;  // probability density value for a given action
   VW::active_multiclass_prediction active_multiclass;
+  char nopred = static_cast<char>(0);
```
Comment (jackgerrits) [inline vowpalwabbit/example.h:73]: To be consistent with label perhaps rename to `no_pred`?

Comment (bassmang) [reply vowpalwabbit/example.h:73]: I changed no_label back to nolabel so its consistent now

**Review by jackgerrits — APPROVED** [review]

**Pull request merged**